README.md
# Secure-Electronic-Voting-System

history.txt
10020384,Wed May  6 10:53:05 2015
10020384,Wed May  6 10:59:37 2015


cli.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <strings.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

int main(int argc, char **argv) {
        int  sockfd, n; 
        char recvline[100], name[10000], ssn[10000], c[10000], vNum[15], c1[10000], okie[5];
	char val[15],voted[15],hist[100];
        struct sockaddr_in servaddr;
	char latRes[100];

	if(argc!=4){
		printf("Please type in this format: ./telnetcli bingsuns.binghamton.edu <portNumber>");
		exit(1); 
	}


        /* Create a TCP socket */
	if((sockfd=socket(AF_INET,SOCK_STREAM, 0)) < 0){
			perror("socket"); exit(2);}
	    

	/* Connect to the server */
	struct hostent *he;
    	struct in_addr **addr_list;
	
	if ((he = gethostbyname(argv[1])) == NULL) {  // get the host info
  	      herror("gethostbyname");
        	return 2;
    	}
//	printf("Official name is: %s\n", he->h_name);
//	printf("    IP addresses: ");
    	addr_list = (struct in_addr **)he->h_addr_list;	
//	printf("%s \n", inet_ntoa(*addr_list[0]));


        /* Specify server's IP address and port */
        bzero(&servaddr, sizeof(servaddr));
        servaddr.sin_family = AF_INET;
        servaddr.sin_port = htons(atoi(argv[2])); /* daytime server port */ 

        if (inet_pton(AF_INET,inet_ntoa(*addr_list[0]) , &servaddr.sin_addr) <= 0) {
			 perror("inet_pton"); exit(3);
        }

       if (connect(sockfd,  (struct sockaddr *) &servaddr,sizeof(servaddr)) < 0 ) {
                 perror("connect");
		exit(4);
	}



	for(;;)
	{
		if(atoi(argv[3]) == 0)
		{


		printf("\n Please Enter your name: >");
		gets(name);

		printf("\n Please enter your Social Security Number:");
		gets(ssn);

		char* old = strcat(name," ");
		char* new = strcat(old, ssn);

		printf("%s\n",new);


		n= write(sockfd, new, 10000);
		
		if (n<0)
		{
			perror("write");
			exit(6);
		} 

		int m = read(sockfd, c, 10000);
		
		
		if(m<0)
		{
			perror("read");
			exit(7);
		}
//		c[m] = '\0';

		printf("%s\n", c);
		
		if(strcmp("NO", c)==0)
		{
			printf("Sorry, you are not eligible to vote.\n");
			exit(10);
		}
		else
		{
			printf("Your validation number = %s\n",c);
			exit(0);
		}			

		exit(0);	
		}//if over


		if(atoi(argv[3]) == 1)
		{
			n= write(sockfd,"1",1);

			read(sockfd,okie,5);

			if(strcmp(okie,"ok")==0)
			{
				printf("\n Enter validation number: >");
				gets(vNum);
			}
			n = write(sockfd, vNum, 15);

			int mi = read(sockfd, c1, 10000);
			
			if(strcmp("Invalid", c1)==0)
			{
				printf("You have entered invalid validation number\n");
				exit(10);
			}
			
			if(strcmp("valid", c1)==0)
			{
			
				while(1)
				{
					int input;
					
					printf("\n\n------------------------------------------------------------------------------------");
					printf("\n Make selection ! :");
					printf("\n 1. Vote");
					printf("\n 2. My vote History");
					printf("\n 3. View the latest results");
					printf("\n 4. Exit");
					printf("\n Enter Your Choice:\t");
					scanf("%d",&input);

					char choice[10];
					int aaa= sprintf(choice, "%d",input);					
			
					write(sockfd,choice,aaa);

					switch(input)
					{
					case 1:
						
						read(sockfd, voted, 15);						
						if(strcmp("voted",voted )==0)
						{
							printf("You have already voted. so you cant vote again");
						}
						else if(strcmp("notvoted",voted )==0)
						{
							int voteNum;
							
							printf("\n\n------------------------------------------------------------------------------------");
							printf("\n Enter number 1 or 2 for vote :");
							printf("\n 1. Bob");
							printf("\n 2. John");
							printf("\n Enter Your Choice:");
							scanf("%d",&voteNum);

							char cho[10];
							int aab= sprintf(cho, "%d",voteNum);					

							n= write(sockfd, cho, aab);

							printf("after write\n");
						}
						break;
							
					case 2:
							
						read(sockfd, hist, 100);
						printf("%s\n",hist);						
						break;
							
					case 3:
						read(sockfd, latRes, 100);
						printf("%s\n", latRes);
						break;
					case 4:
						exit(0);
					default:
						break;								
					}
					
				//	write(sockfd, input, 10);
					}
			}
		}


	
	}
	close(sockfd);	
} 



laser.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <strings.h>
#include <string.h>
#include<time.h>
#include<netdb.h>


void socket_program (int port, int ssn_bit);

int main(int argc, char **argv) {
    int   listenfd, connfd, clilen,ssnbit;
    struct sockaddr_in servaddr, cliaddr;
    char buff[10000];
	char *name,*ftemp, *ssn, *ftemp2;
	char line[100],fname[40],fssn[25],fstatus[50];

    time_t ticks;
	srand (time(NULL));

	if(argc!=3)
	{
        printf("Please type in this format: ./telnetserv <portNumber>");
        exit(1);
    }


    /* Create a TCP socket */
    listenfd = socket(AF_INET, SOCK_STREAM, 0);

	/* Initialize server's address and well-known port */
	bzero(&servaddr, sizeof(servaddr));
    servaddr.sin_family      = AF_INET;
    servaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    servaddr.sin_port        = htons(atoi(argv[1]));   /* daytime server */

	/* Bind servers address and port to the socket */
    bind(listenfd, (struct sockaddr *) &servaddr, sizeof(servaddr));       
	/* Convert socket to a listening socket max 100 pending clients*/
	listen(listenfd, 100); 

for(;;)
{

   
	clilen = sizeof(cliaddr);
	connfd = accept(listenfd, (struct sockaddr *)&cliaddr, &clilen);

	int n ;
	n = read(connfd, buff, 10000);
	if(n<0)
	{
		perror("read");
		exit(5);
	}
	char *search = " ";

	name = strtok(buff, search);
	ssn = strtok(NULL, search);
	int flag=0;

	printf("Name: %s\n",name);
	printf("SSN: %s\n",ssn);
	
	FILE *fl;
	fl = fopen ("status.txt", "rt");
		
	while(fgets(line, 100, fl) != NULL)
   	{
		ftemp = strtok(line,",");
		int i =0;

		while(ftemp != NULL)
		{
			if(i == 0)
			{
				strcpy(fname,ftemp);		
			}
			else if(i==1)
			{
				strcpy(fssn,ftemp);
			}
			else
			{
				strcpy(fstatus,ftemp);		
			}

			ftemp = strtok(NULL,",");
			i++;
		}
		printf("%s\n",fname);
		printf("%s\n",fssn);
		printf("%s\n",fstatus);
    
		if (strcmp(fname,name)==0 && strcmp(fssn,ssn)==0)
		{
			printf("his status =%s\n",fstatus);
			if(strcmp(fstatus,"citizen\n")==0)
			{
				int mainflag = 0;
				FILE *f22;
				f22 = fopen ("verify.txt", "rb");
				char line22[100];
				char *v2,*l2;
				while(fgets(line22, 100, f22) != NULL)
				{
				
					v2 = strtok(line22,",");
					l2 = strtok(NULL," \n");
					if(strcmp(v2, ssn)==0)
					{
						if(strcmp("0",l2)==0)
						{
							mainflag = 1;
							break;
						}
						else
						{
							write(connfd,l2,50);
							break;
						}
					}
				}
				fclose(f22);
				

				if(mainflag == 1)
				{
			
					FILE *f2;
					f2 = fopen("verify.txt", "rb");
					char line2[100];

					printf("%d\n",mainflag);

					FILE *ftemp;
					ftemp = fopen ("temp.txt","w+");

					//flag =1 ;
					char *v1,*l1;


					ssnbit = rand() % 89999999 + 10000000;
					printf("%d\n",ssnbit);
					
					char ssnb[15];
					int ss = sprintf(ssnb,"%d",ssnbit);
					write(connfd,ssnb,15);	

					while(fgets(line2, 100, f2) != NULL)
					{
				
						v1 = strtok(line2,",");
						l1 = strtok(NULL," ");


						if(strncmp(fssn,v1,9)==0)
						{
							fprintf(ftemp,"%s,%d\n",v1,ssnbit);
						
						}
						else
						{
							fprintf(ftemp,"%s,%s",v1,l1);					
						}

					}
					fclose(ftemp);
					fclose(f2);



					if(flag == 0)
					{
						int ret = remove("verify.txt");
						int ret2 = rename("temp.txt","verify.txt");
					}
			printf("socket program k upar");
					socket_program(atoi(argv[2]),ssnbit);


				}

			}
			else
			{
				flag =1;
				int mm= write(connfd,"NO",strlen("NO"));
			}
		}
		
	}
	fclose(fl);
		
	printf("flag value=%d\n",flag);
	

		bzero(buff, 10000);	
		printf("%s", buff);

}


    close(connfd);
    close(listenfd);

//	socket_program(atoi(argv[1]),ssnbit);

}








void socket_program (int port, int ssn_bit)
{
	int  sockfd;
	char recvline[100] ,reply[10];
	struct sockaddr_in servaddr;

	/* Create a TCP socket */
	if((sockfd=socket(AF_INET,SOCK_STREAM, 0)) < 0){
	perror("socket"); exit(2);}
	    
	/* Connect to the server */
	struct hostent *he;
	struct in_addr **addr_list;

	
	if ((he = gethostbyname("bingsuns.binghamton.edu"))== NULL)
	{  // get the host info
		herror("gethostbyname");
		//return 2;
		exit(0);
	}


//	gethostbyname("bingsuns.binghamton.edu");


	//	printf("Official name is: %s\n", he->h_name);
	//	printf("    IP addresses: ");


	addr_list = (struct in_addr **)he->h_addr_list;	
	//	printf("%s \n", inet_ntoa(*addr_list[0]));
	/* Specify server's IP address and port */
	bzero(&servaddr, sizeof(servaddr));
	servaddr.sin_family = AF_INET;
	servaddr.sin_port = htons(port); /* daytime server port */ 
	if (inet_pton(AF_INET,inet_ntoa(*addr_list[0]) , &servaddr.sin_addr) <= 0)
	{
		perror("inet_pton"); exit(3);
	}
	if (connect(sockfd,  (struct sockaddr *) &servaddr,sizeof(servaddr)) < 0 )
	{
		perror("connect");
		exit(4);
	}

	printf("in socket code");

	int op= write(sockfd,"0",strlen("0"));
	printf("write 0 hua\n");

	int n = read(sockfd, reply, 10);
	printf("ok read hua\n");

	if(strcmp(reply,"ok")==0)
	{

		printf("if condition me gaya\n");

		char ssb[15];
		int sb = sprintf(ssb,"%d",ssn_bit);
		printf("ssb:%s, sb:%d, ssn_bit:%d\n ",ssb,sb,ssn_bit);

		int wr = write(sockfd,ssb,strlen(ssb));
		if(wr <0)
			printf("problem happened in write");	

	}
}


vfser.c

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <strings.h>
#include <string.h>
#include<time.h>

void voting(char* validNo);
void history(char* validNo);
void result();
char buff[10000],buff2[8];
char *validNo, *yesNo, *histo, *name3, *calc, *validNo3, *v1,*l1,*v2,*l2;

char line[100],line2[100],line3[100],line4[100],line5[100],line6[100],line7[100],line8[100],line9[100], choice[10],selection[1],vote[5];
int listenfd, connfd, clilen, input;
int flag=0;
struct sockaddr_in servaddr, cliaddr;


int main(int argc, char **argv) {

    time_t ticks;
	srand (time(NULL));

	if(argc!=2)
	{
        printf("Please type in this format: ./telnetserv <portNumber>");
        exit(1);
    }


    /* Create a TCP socket */
    listenfd = socket(AF_INET, SOCK_STREAM, 0);

	/* Initialize server's address and well-known port */
	bzero(&servaddr, sizeof(servaddr));
    servaddr.sin_family      = AF_INET;
    servaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    servaddr.sin_port        = htons(atoi(argv[1]));   /* daytime server */

	/* Bind servers address and port to the socket */
    bind(listenfd, (struct sockaddr *) &servaddr, sizeof(servaddr));       
	/* Convert socket to a listening socket max 100 pending clients*/
	listen(listenfd, 100); 

	for(;;)
	{

   
		clilen = sizeof(cliaddr);
	        connfd = accept(listenfd, (struct sockaddr *)&cliaddr, &clilen);

		int n ;
		int uv = read(connfd, selection, 1);
		
		if(strcmp(selection,"0")==0)
		{
			flag = 2;
		}
	
		if(strcmp(selection,"1")==0)
		{
			flag = 3;
		}
	
		printf("selection:a%sa\n",selection);


		//n = read(connfd, buff, 10000);

		if(n<0)
		{
			perror("read");
			exit(5);
		}

		int flagbit = 0;
//		printf("voternumber in buffer:%s\n",buff);

		if(flag ==2)
		{
			printf("flag = %d\n",flag);
			write(connfd,"ok",10);
			printf("after write and before read in if\n");

			int prob = read(connfd, buff2, 8);
			if(prob < 0)
			{
				printf("problem with reading voter number\n");
			}

			printf("validation number read hua");

			FILE *f1;
			f1 = fopen ("voternumber.txt", "a+");
		
			//fseek(f1, -100, SEEK_END);
		    	fprintf(f1,"%s,0\n",buff2);
			//exit(5);
			printf("vallNum:%s, done with procesing LA",buff2);
			fclose(f1);
		
		}
		else if( flag == 3)
		{
			write(connfd,"ok",5);


			n = read(connfd, buff, 10000);


			FILE *f2;
			f2 = fopen ("voternumber.txt", "rw+");

		
			while(fgets(line2, 100, f2) != NULL)
			{
				char *search = ",";
				validNo = strtok(line2, search);
				yesNo = strtok(NULL, search);
			
				printf("validation number: %s\n",validNo);				
				printf("Yes/No: %s\n",yesNo);
				printf("buffer:%s",buff);	


				if(strcmp(buff,validNo)==0)
				{
					flagbit = 1;
					break;
				}
			}	
			
			if(flagbit == 1)
			{
				int m= write(connfd,"valid",strlen("valid"));

				n = read(connfd, choice, 10);

				input = atoi(choice);

				if(input==4)
				exit(0);

				switch(input)
		     		{
		   			case 1:
					printf("Voting is Selected :)\n");
					voting(buff);
					break;
	
				    case 2:
					printf("My vote history is Selected :)\n");
					history(buff);
					break;
	
					case 3:
					printf("View the latest result is Selected :)\n");
					result();
					break;
	
					case 4:
					exit(0);
		
					default:
					break;
		     		}

			}
			else
			{
				int m= write(connfd,"Invalid",strlen("Invalid"));
			}
		}


		/*	int m = write(connfd,buff, strlen(buff));
		
		if(m<0){
		perror("write error");
		exit(5);
		} 
		*/

		bzero(buff, 10000);	
		printf("%s", buff);


	}


    close(connfd);
	close(listenfd);
}

void voting(char* validNo)
{
	char line3[100];
	char* validNo2, *hi;
	int flagbit2=0, abc;

	FILE *f3;
	f3 = fopen ("voternumber.txt", "r");
	

	printf("inside voting\n");
		
	while(fgets(line3, 100, f3) != NULL)
   	{
		printf("inside while\n");
		char *search = ",";
		validNo2 = strtok(line3, search);
		hi = strtok(NULL, search);

		printf("validation number: %s\n",validNo2);
		printf("voted or not: a%sa\n",hi);
	
		
		if(strcmp(validNo,validNo2)==0)
		{
			if(strcmp(hi,"0\n")==0)
			{
				flagbit2 == 1;
				break;
			}
		}
    		
	}
	fclose(f3);

	printf("flagbit2: %d", flagbit2);

	if(flagbit2 == 0)
	{
		int p= write(connfd,"voted",strlen("voted"));
	}
	else
	{
		int k= write(connfd,"notvoted",strlen("notvoted"));

		int n = read(connfd, vote, 10);


		printf("vote: %s",vote);

		FILE *f5;
		f5 = fopen ("result.txt", "r");
		int e=0;
		char l11[15],l12[15],te1[5],te2[5];

		while(fgets(line5, 100, f5) != NULL)
		{	
			if (e ==0)
				{
					v1 = strtok(line5,",");
					l1 = strtok(NULL," ");
					printf("e =0\n");
					printf("v1:%s\n",v1);
					printf("l1:%s\n",l1);

					int x = atoi(l1);
					if(strcmp(vote,"1")==0)
					{
						x++;
					}
//					l1= itoa(x);
					int bv = sprintf(te1,"%d",x);			

					strcpy(l11,v1);
					strcat(l11,",");
					strcat(l11,te1);
				}
			if (e==1)
				{
					v2 = strtok(line5,",");
					l2 = strtok(NULL," ");
					printf("e =1\n");
					printf("v2:%s\n",v2);
					printf("l2:%s\n",l2);
							
					int y = atoi(l2);
					if(strcmp(vote,"2")==0)
					{
   						y++;
					}
//					l2= itoa(y);
					int bw = sprintf(te2,"%d",y);					
					strcpy(l12,v2);
					strcat(l12,",");
					strcat(l12,te2);

				}
				e++;
		}
		//fclose(f5);


//					printf("outside v1:%s\n",v1);
//					printf("outside l1:%s\n",l1);
//					printf("outside v2:%s\n",v2);
//					printf("outside l2:%s\n",l2);

	
					printf("outside l11:%s\n",l11);
					printf("outside l12:%s\n",l12);


		FILE *f6;
		f6 = fopen ("temp1.txt","w+");
		if(f6 == NULL){printf("File error");}
//		if(strcmp(vote,"1")==0)
//		{
//			fprintf(f6,"%s,%d\n",v1,atoi(l1));
//			fprintf(f6,"%s,%s\n",v2,l2);
						
//		}
//		if(strcmp(vote,"2")==0)
//		{
			fprintf(f6,"%s\n",l11);
			fprintf(f6,"%s\n",l12);
//		}

		fclose(f6);
		fclose(f5);		
			
		int r1 = remove("result.txt");
		int r2 = rename("temp1.txt","result.txt");
			

		FILE *f4;
		f4 = fopen ("history.txt", "a");

		time_t curtime;
   		time(&curtime);
		fprintf(f4,"%s,%s",validNo,ctime(&curtime));

		printf("end tak pahucha\n");

		fclose(f4);
		

			
					FILE *f200;
					f200 = fopen("voternumber.txt", "rb");
					char line200[100];

					FILE *ftemp100;
					ftemp100 = fopen ("temp100.txt","w+");
					
					//flag =1 ;
					char *v200,*l200;


					while(fgets(line200, 100, f200) != NULL)
					{
				
						v200 = strtok(line200,",");
						l200 = strtok(NULL," ");

					if(strcmp(validNo,v200)==0)
							fprintf(ftemp100,"%s,1\n",v200);
						
					else
							fprintf(ftemp100,"%s,%s\n",v200,l200);


					}
					fclose(ftemp100);
					fclose(f200);

					remove("voternumber.txt");
					rename("temp100.txt","voternumber.txt");
								
								
		
		
		
		
		
	}
}



void history(char* validNo)
{

	char hist[100];
	FILE *f7;
	f7 = fopen("history.txt", "r+");
		
	while(fgets(line3, 100, f7) != NULL)
   	{
		char *search = ",";
		validNo3 = strtok(line3, search);
		histo = strtok(NULL, search);

		printf("validation number: %s\n",validNo3);
		printf("History: %s\n",histo);
	
		if(strcmp(validNo,validNo3)==0)
		{
			strcpy(hist,histo);
			printf("history:%s",hist);
			int o= write(connfd,hist,100);
			break;
		}		
	}
	fclose(f7);
}


void result()
{
	char resu[100];
	FILE *f8;
	f8 = fopen("result.txt", "r+");
		
	while(fgets(line8, 100, f8) != NULL)
   	{
		strcat(resu,line8);
	}
	int o= write(connfd,resu,100);	
	fclose(f8);

}
